### PR TITLE
Add Nonce and N Checking

### PR DESCRIPTION
**Pull request by hayesgm**
This patch adds support for setting the active nonce and submission token in temporary storage so scripts can read it and get the current nonce, submission token, and if they so choose, N. This is useful for certain scripts.

### CODE REVIEW

**Comment by kevincheng96** [comment]
LGTM on the changes thus far. Will give approval after tests

File: src/quark-core/src/QuarkScript.sol [hunk 1]
```diff
@@ -118,4 +120,54 @@ abstract contract QuarkScript {
...
+
+    // Note: this may not be accurate after any nested calls from a script
+    function getActiveNonce() internal returns (bytes32) {
+        QuarkWallet self = QuarkWallet(payable(address(this)));
+
+        bytes32 activeNonceSlot = self.ACTIVE_NONCE_SLOT();
+        bytes32 value;
+        assembly {
+            value := sload(activeNonceSlot)
+        }
```
Comment (fluffywaffles) [inline src/quark-core/src/QuarkScript.sol:132]: do we not have `read()`/`write()` helpers that handle the `bytes32` case? would rather avoid `assembly` blocks and `sload`s if we can

File: src/quark-core/src/QuarkScript.sol [hunk 2]
```diff
@@ -118,4 +120,54 @@ abstract contract QuarkScript {
...
+
+    // Note: this may not be accurate after any nested calls from a script
+    // Returns the active replay count of this script. Thus, the first submission should return 0,
+    // the second submission 1, and so on. This must be called before the script makes any external calls.
+    function getActiveReplayCount() internal returns (uint256) {
+        bytes32 nonce = getActiveNonce();
+        bytes32 submissionToken = getActiveSubmissionToken();
+        uint256 n;
+
+        if (submissionToken == bytes32(type(uint256).max)) {
```
Comment (fluffywaffles) [inline src/quark-core/src/QuarkScript.sol:163]: can we not do `QuarkNonceManager.EXHAUSTED`?

the tiny workaround here is you can do what we did in `QuarkWallet` with the `QuarkWalletMetadata` library:

```solidity
library SpecialNonce {
  bytes32 internal constant FREE = bytes32(0);
  bytes32 internal constant EXHAUSTED = bytes32(type(uint256).max);
}
```

and then `SpecialNonce.EXHAUSTED` gets inlined everywhere. 🤷

File: test/lib/CancelOtherScript.sol [hunk 3]
```diff
@@ -2,11 +2,24 @@
 pragma solidity 0.8.23;
 
 import "quark-core/src/QuarkWallet.sol";
+import "quark-core/src/QuarkScript.sol";
 
-contract CancelOtherScript {
+contract CancelOtherScript is QuarkScript {
```
Comment (fluffywaffles) [inline test/lib/CancelOtherScript.sol:7]: does this contract actually make any sense anymore? I don't think it really does

you could rename it `Cancel` but it's at least no longer a `CancelOther`, and really it's a `nop` that you could just as easily run to emit the `CancelNonce()` event erroneously...

**Review by fluffywaffles — APPROVED** [review]
overall, looks good

File: src/quark-core/src/QuarkWallet.sol [hunk 4]
```diff
@@ -113,6 +117,13 @@ contract QuarkWallet is IERC1271 {
     /// @notice Well-known storage slot for the currently executing script's address (if any)
     bytes32 public constant ACTIVE_SCRIPT_SLOT = bytes32(uint256(keccak256("quark.v1.active.script")) - 1);
 
+    /// @notice Well-known --
+    bytes32 public constant ACTIVE_NONCE_SLOT = bytes32(uint256(keccak256("quark.v1.active.nonce")) - 1);
+
+    /// @notice Well-known --
+    bytes32 public constant ACTIVE_SUBMISSION_TOKEN_SLOT =
+        bytes32(uint256(keccak256("quark.v1.active.submissionToken")) - 1);
```
[the remark this answers is not in the document]
Comment (hayesgm) [reply src/quark-core/src/QuarkWallet.sol:125]: fixed in #210

File: src/quark-core/src/QuarkScript.sol (hunk 1 above)
Comment (hayesgm) [reply src/quark-core/src/QuarkScript.sol:140]: I think those are about to get isolation, so probably doesn't work?

**Pull request merged**